### PR TITLE
Fix Angular externs to work with TS 1.4

### PR DESCRIPTION
**Pull request by mprobst**
Merge the IAngularStatic and the angular namespace by moving all type definitions into the main "angular" module.
This fixes compatibility with TypeScript 1.4 (see issue #3656 and issue #3670), and allow depending on the definitions using "require" style imports instead of <reference> style.

The drawback is that this is a compatibility breaking change. Existing clients must replace all occurences of "ng." in their code base with "angular.", for example by running `sed -i 's|\bng\.|angular.|g'` or something similar over their entire code base.

I don't see another reasonable way to create compatibility with TS 1.4 - any other ideas that'd maintain backwards compatibility? Maybe it'd be possible to declare a `var ng = angular;` in a compatibility shim?

### CODE REVIEW

**Comment by xi** [comment]
I pass the angular "module" to some function in my code. Without this, I can use the interface `ng.IAngularStatic` for that. How would you do this with this new version?

**Comment by mprobst** [comment]
Good point. Can you just refer to the module as a type, i.e. `function foo(ng: angular) { ... }`?

Also: what's the use case for this? Are you mocking Angular itself? Why?

**Comment by xi** [comment]
I pass angular via dependency injection in order to be able to mock it and run my unit tests without a browser.

**Comment by mprobst** [comment]
I talked with @mhegazy about this, I think my definitions are the more TypeScriptian way to do this, and compatible with the external module approach.

@borisyankov is there interest in this? Should I update the pull request?

File: angularjs/angular.d.ts [hunk 1]
```diff
@@ -1325,13 +1313,13 @@ declare module ng {
...
     // TemplateCacheService
-    // see http://docs.angularjs.org/api/ng.$templateCache
+    // see http://docs.angularjs.org/api/angular.$templateCache
     ///////////////////////////////////////////////////////////////////////////
     interface ITemplateCacheService extends ICacheObject {}
 
     ///////////////////////////////////////////////////////////////////////////
     // SCEService
-    // see http://docs.angularjs.org/api/ng.$sce
+    // see http://docs.angularjs.org/api/angular.$sce
```
Comment (mhegazy) [inline angularjs/angular.d.ts]: I do not think these are correct. find and replace issue on my side

**Comment by mhegazy** [comment]
for IAngularStatic, you can just define a new type alias:
```ts
declare module angular {
    type IAngularStatic = typeof angular;
    ...
}
```

this will keep the original behavior working as it was.

File: angularjs/angular.d.ts (hunk 1 above)
Comment (mprobst) [inline angularjs/angular.d.ts]: Actually, that was my own find and replace that I botched here :-)

Good catch.

**Comment by mprobst** [comment]
@mhegazy is it possible to alias the type `angular` to `ng` for backwards compatibility with existing users? I tried something along these lines:

```js
/// <reference path="angular.d.ts" />
declare type ng = typeof angular;
// To use ng like this:
var myAngular: ng.IModule = angular.module('asd', []);
```

But I get this error:
```
ng-tests.ts(3,16): error TS2304: Cannot find name 'ng'.
```

Which I guess means you cannot reference members of type aliases?

**Comment by mhegazy** [comment]
How about:

```ts
import ng = angular;
```

Consumers (e.g. tests in this case) can use add that, or you can put it in the .d.ts file; but if you put this in the .d.ts file, that means you have an alias "ng" to "angular" that just has the same meaning in the global namespace. up to you i guess.

**Comment by mprobst** [comment]
Thanks, that works.

**Comment by vvakame** [comment]
sorry I can't understand why this change is required.
this PR has a breaking change that it is very heavy.
what do you think about? @basarat

**Comment by basarat** [comment]
@mprobst I believe this PR is to allow the user to consistently use 'angular' everywhere : 

```ts
import angular = require('angular');
var myMod: angular.Module = angular.module(...);
```
Is that the only motivation? I

**Comment by mprobst** [comment]
@basarat indeed, my goal is to have these declarations work both when using them as internal and when using them as external modules. Admittedly, I have no idea what I'm doing. I have tests for both, any clue on how to make that work?

**Pull request closed without merging**